### PR TITLE
Add CUDA support for D3D11 interop. 

### LINKED ISSUE #13888: OpenCL: Can't create context for DirectX interop in function cv::directx::ocl::initializeContextFromD3D11Device
##### System information (version)

- OpenCV =>                                   4.0 (build with contrib and cuda)
- Operating System / Platform =>  Win10 x64
- Compiler =>                                  VS2017:
- CUDA=>                                        10.0
- GPU=>                                           GTX 1050Ti
- GPU Device=>                               417.35
- OpenCL=>                                     1.2

##### Detailed description
`cv::directx::ocl::initializeContextFromD3D11Device();`
`cv::directx::convertFromD3D11Texture2D()`

There are the same problem：

`Exceptions thrown at 0x0000000000000 (in .exe): Access conflicts occur when 0xC0000005: Execution location 0x0000000000000000.`

##### Steps to reproduce

```cpp
void createDevice(ID3D11Device *&d3dDevice) {
	UINT createDeviceFlags = 0;

#ifdef _DEBUG
	createDeviceFlags |= D3D11_CREATE_DEVICE_DEBUG;
#endif

	D3D_FEATURE_LEVEL featureLevel;
	ID3D11DeviceContext*  md3dImmediateContext;
	HRESULT  hr = D3D11CreateDevice(
		0,                      
		D3D_DRIVER_TYPE_HARDWARE,
		0,                      
		createDeviceFlags,
		0,                     
		0,                     
		D3D11_SDK_VERSION,    
		&d3dDevice,         
		&featureLevel, 
		&md3dImmediateContext);

	if (FAILED(hr))
	{
		OutputDebugStringA("D3D11CreateDevice Failed.");
	}
	if (featureLevel != D3D_FEATURE_LEVEL_11_0)
	{
		OutputDebugStringA("Direct3D FeatureLevel 11 unsupported.");
	}
}

int main() {
	ID3D11Device* d3dDevice=nullptr;
	createDevice(d3dDevice);
        cv::ocl::Context oclContext=cv::directx::ocl::initializeContextFromD3D11Device(d3dDevice);
}
```

### PR DESCRIPTION
**Pull request by Mainvooid**
resolves #13888

### This pullrequest changes

Add CUDA support(cl_nv_d3d11_sharing) for D3D11 interop.

### CODE REVIEW

File: modules/core/src/directx.cpp [hunk 1]
```diff
@@ -248,25 +248,106 @@ Context& initializeContextFromD3D11Device(ID3D11Device* pD3D11Device)
...
-
-    int found = -1;
-    cl_device_id device = NULL;
-    cl_uint numDevices = 0;
-    cl_context context = NULL;
-
+	cl_uint numPlatforms;
+	cl_int status = clGetPlatformIDs(0, NULL, &numPlatforms);
+	if (status != CL_SUCCESS)
+		CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get number of platforms");
```
Comment (alalek) [inline modules/core/src/directx.cpp]: Revert indentation changes.
Don't use 'tabs'.

File: 3rdparty/include/opencl/1.2/CL/cl_d3d11_ext.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+ **********************************************************************************/
+
+#ifndef __OPENCL_CL_D3D11_EXT_H
+#define __OPENCL_CL_D3D11_EXT_H
```
Comment (alalek) [inline 3rdparty/include/opencl/1.2/CL/cl_d3d11_ext.h]: Please add information about source of this file.

File: modules/core/src/directx.cpp [hunk 3]
```diff
@@ -248,25 +248,106 @@ Context& initializeContextFromD3D11Device(ID3D11Device* pD3D11Device)
...
+
+	// TODO Filter platforms by name from OPENCV_OPENCL_DEVICE.
+	int found = -1;
+	cl_device_id device = NULL;
+	cl_uint numDevices = 0;
+	cl_context context = NULL;
+
+#ifdef HAVE_CUDA
```
Comment (alalek) [inline modules/core/src/directx.cpp]: Do we really need CUDA SDK for this?
I believe it is "OpenCL" feature and CUDA is not used here.

File: modules/core/src/directx.cpp [hunk 4]
```diff
@@ -678,43 +759,65 @@ Context& initializeContextFromDirect3DDevice9(IDirect3DDevice9* pDirect3DDevice9
...
+	}
+#endif// HAVE_CUDA
 }
-#endif // defined(HAVE_DIRECTX) && defined(HAVE_OPENCL)
+#endif// defined(HAVE_DIRECTX) &&  defined(HAVE_OPENCL)
 
 } // namespace directx
 
-
 namespace ocl {
```
Comment (alalek) [inline modules/core/src/directx.cpp]: Please avoid unrelated changes.

File: modules/core/src/directx.cpp [hunk 5]
```diff
@@ -678,43 +759,65 @@ Context& initializeContextFromDirect3DDevice9(IDirect3DDevice9* pDirect3DDevice9
...
 
 #if defined(HAVE_DIRECTX) && defined(HAVE_OPENCL)
 #ifdef HAVE_DIRECTX_NV12
 
-static
-bool ocl_convert_nv12_to_bgr(
+static bool ocl_convert_nv12_to_bgr(
     cl_mem clImageY,
```
Comment (alalek) [inline modules/core/src/directx.cpp]: Please avoid unrelated changes.

File: modules/core/src/directx.cpp (hunk 3 above)
Comment (Mainvooid) [reply modules/core/src/directx.cpp]: The current situation is that the NVIDIA CUDA platform does not support the `cv_khr_d3d11_sharing` extension, so in order to support D3D11 interoperability on NVIDIA CUDA platform, we need to support the `cv_nv_d3d11_sharing` extension. Current changes ensure that users whose compilation options are selected for WITH_CUDA can use the NVIDIA CUDA platform, while unchecked ones do not affect the original functionality. Perhaps it would be a better way to determine whether an OCL device has a `cv_nv_d3d11_sharing` extension? But at the moment I'm not sure how to judge this at compile time. It would be great if you could give me some advice.

File: modules/core/src/directx.cpp (hunk 1 above)
Comment (Mainvooid) [reply modules/core/src/directx.cpp]: Repaired

File: 3rdparty/include/opencl/1.2/CL/cl_d3d11_ext.h (hunk 2 above)
Comment (Mainvooid) [reply 3rdparty/include/opencl/1.2/CL/cl_d3d11_ext.h]: Information has been added to the file.

File: modules/core/src/directx.cpp (hunk 5 above)
Comment (Mainvooid) [reply modules/core/src/directx.cpp]: It has been restored, although there should be no difference between them.

File: modules/core/src/directx.cpp (hunk 4 above)
Comment (Mainvooid) [reply modules/core/src/directx.cpp]: Just delete blank lines.

File: modules/core/src/directx.cpp (hunk 3 above)
Comment (Mainvooid) [reply modules/core/src/directx.cpp]: What if I change it like this?
[`OpenCVDetectOpenCL.cmake`](https://github.com/opencv/opencv/blob/master/cmake/OpenCVDetectOpenCL.cmake):

```cpp
if(OPENCL_FOUND)
...
  try_compile(__VALID_CL_KHR_D3D11_SHARING
    "${OpenCV_BINARY_DIR}"
    "${OpenCV_SOURCE_DIR}/cmake/checks/opencl.cpp"
    COMPILE_DEFINITIONS "-DCHECK_CL_KHR_D3D11_SHARING"
    OUTPUT_VARIABLE TRY_OUT
  )
  if(__VALID_CL_KHR_D3D11_SHARING)
    set(HAVE_CL_KHR_D3D11_SHARING ON)
  endif()

  try_compile(__VALID_CL_NV_D3D11_SHARING
    "${OpenCV_BINARY_DIR}"
    "${OpenCV_SOURCE_DIR}/cmake/checks/opencl.cpp"
    COMPILE_DEFINITIONS "-DCHECK_CL_NV_D3D11_SHARING"
    OUTPUT_VARIABLE TRY_OUT
  )
  if(__VALID_CL_NV_D3D11_SHARING)
    set(HAVE_CL_NV_D3D11_SHARING ON)
  endif()
...
endif()
```

[`cmake/check/opencl.cpp`](https://github.com/opencv/opencv/blob/master/cmake/checks/opencl.cpp):

```cpp
#include <opencv2/core/ocl.hpp>
...
int main(int /*argc*/, char** /*argv*/)
{
#ifdef CHECK_CL_KHR_D3D11_SHARING
    bool khr_d3d11 = cv::ocl::Device::getDefault().isExtensionSupported("cl_khr_d3d11_sharing");
#endif

#ifdef CHECK_CL_NV_D3D11_SHARING
    bool nv_d3d11 = cv::ocl::Device::getDefault().isExtensionSupported("cl_nv_d3d11_sharing");
#endif
    return 0;
}
```

File: cmake/checks/opencl.cpp [hunk 6]
```diff
@@ -17,5 +20,19 @@
 
 int main(int /*argc*/, char** /*argv*/)
 {
+#ifdef CHECK_CL_KHR_D3D11_SHARING
+    bool khr_d3d11 = cv::ocl::Device::getDefault().isExtensionSupported("cl_khr_d3d11_sharing");
```
Comment (Mainvooid) [inline cmake/checks/opencl.cpp]: Should it be judged within a function rather than at compile time?

File: modules/core/src/directx.cpp (hunk 3 above)
Comment (Mainvooid) [reply modules/core/src/directx.cpp]: The build log display buildbot also does not support the `cl_khr_d3d11_sharing` extension.

**Comment by alalek** [comment]
1.
We need compile time check for having `cl_d3d11_ext.h`.
OpenCV can be build with external OpenCL SDK (useful for Android builds) without this file with NV extensions.

2.
Detect OpenCV device extension "cl_nv_d3d11_sharing"/etc in runtime.
Don't use "Device::getDefault()" here - this call initializes some OpenCL device (by OpenCV choice). You need to select your own - use `Device(void* d /* cl_device_id d */)` constructor.

File: modules/core/src/directx.cpp (hunk 3 above)
Comment (alalek) [reply modules/core/src/directx.cpp]: `cl_khr_d3d11_sharing` is available with AMD devices (probably on Windows). Currently we don't have such build workers.

**Comment by Mainvooid** [comment]
> Detect OpenCV device extension "cl_nv_d3d11_sharing"/etc in runtime.
> Don't use "Device::getDefault()" here - this call initializes some OpenCL device (by OpenCV choice). You need to select your own - use `Device(void* d /* cl_device_id d */)` constructor.

There is a problem ,the  `Device(void* d/*cl_device_id d*/)` object is NULL before we  call `clGetDeviceIDsFromD3D11NV`or`clGetDeviceIDsFromD3D11KHR` ,that is what we need to early determination.

Do you have any good suggestions to get the OCL device information object from ID3D11Device to support us to dynamically call the corresponding functions according to the current device?

**Comment by Mainvooid** [comment]
> Detect OpenCV device extension "cl_nv_d3d11_sharing"/etc in runtime.
> Don't use "Device::getDefault()" here - this call initializes some OpenCL device (by OpenCV choice). You need to select your own - use `Device(void* d /* cl_device_id d */)` constructor.

There is a problem ,the  `Device(void* d/*cl_device_id d*/)` object is NULL before we  call `clGetDeviceIDsFromD3D11NV`or`clGetDeviceIDsFromD3D11KHR` ,that is what we need to early determination.

Do you have any good suggestions to get the OCL device information object from `ID3D11Device` to support us to dynamically call the corresponding functions according to the current `ID3D11Device's` own OCL device? If we can't get OCL information from D3D, do we need to give the platform to users to set up?

Or is it just a matter of deciding whether the first platform supports corresponding extensions?
```cpp
cl_platform_id platform = (cl_platform_id) Platform:: getDefault().PTR();
size_t size;
// extensions
cl_uint status= clGetPlatformInfo(platform, CL_PLATFORM_EXTENSIONS, 0, NULL, &size);
cv::AutoBuffer<char> extensions(size);
status= clGetPlatformInfo(platform, CL_PLATFORM_EXTENSIONS, size, extensions, NULL);
if (strstr(extensions, "cl_nv_d3d11_sharing")) {
    printf("cl_nv_d3d11_sharing is supported");
}
```

**Comment by alalek** [comment]
> "cl_nv_d3d11_sharing"

I believe we could check result of `clGetExtensionFunctionAddressForPlatform(platform, "clCreateFromD3D11Texture2DNV");` for NULL instead of checking extensions.

**Comment by Mainvooid** [comment]
If there are no problems, I may provide further support for D3D9 and D3D10 later.

**Comment by Mainvooid** [comment]
`HAVE_CLNVEXT` is not turned on by default. How do I handle unreferenced local functions and formal parameters?

File: modules/core/src/directx.cpp [hunk 7]
```diff
@@ -260,35 +264,35 @@ Context& initializeContextFromD3D11Device(ID3D11Device* pD3D11Device)
...
     cl_context context = NULL;
 
-    // try with CL_PREFERRED_DEVICES_FOR_D3D11_KHR
+#ifdef HAVE_CLNVEXT
+    // try with CL_PREFERRED_DEVICES_FOR_D3D11_NV(Need to check it first)
     for (int i = 0; i < (int)numPlatforms; i++)
     {
-        clGetDeviceIDsFromD3D11KHR_fn clGetDeviceIDsFromD3D11KHR = (clGetDeviceIDsFromD3D11KHR_fn)
-                clGetExtensionFunctionAddressForPlatform(platforms[i], "clGetDeviceIDsFromD3D11KHR");
-        if (!clGetDeviceIDsFromD3D11KHR)
```
Comment (alalek) [inline modules/core/src/directx.cpp]: Non-NVIDIA code should be restored.

File: cmake/OpenCVDetectOpenCL.cmake [hunk 8]
```diff
@@ -77,4 +77,21 @@ if(OPENCL_FOUND)
...
+      NAMES CL/cl_d3d11_ext.h
+      PATHS ENV CUDA_PATH ENV ProgramFiles
+      PATH_SUFFIXES include
+      DOC "clNvExt include directory")
+
+    if(CLNVEXT_INCLUDE_DIR)
+      set(HAVE_CLNVEXT on)
+      list(FIND OPENCL_INCLUDE_DIRS "${CLNVEXT_INCLUDE_DIR}" FIND_OUT)
+      if (FIND_OUT MATCHES -1)
+        list(APPEND OPENCL_INCLUDE_DIRS "${CLNVEXT_INCLUDE_DIR}")
```
Comment (alalek) [inline cmake/OpenCVDetectOpenCL.cmake]: Perhaps this would introduce mess with OpenCL headers from **multiple** include directories.

Check `OPENCL_INCLUDE_DIR` path only for `cl_d3d11_ext.h` header.

So your usage should be:
```
-DOPENCL_INCLUDE_DIR=<CUDA_PATH>/include
-DWITH_OPENCL_D3D11_NV=ON
```

File: cmake/templates/cvconfig.h.in [hunk 9]
```diff
@@ -100,6 +100,9 @@
 #cmakedefine HAVE_OPENCL_STATIC
 #cmakedefine HAVE_OPENCL_SVM
 
+/* NVIDIA OpenCL Extensions support */
+#cmakedefine HAVE_CLNVEXT
```
Comment (alalek) [inline cmake/templates/cvconfig.h.in]: Please use more readable name:
```
HAVE_OPENCL_NV_EXT
```

or more concrete:
```
HAVE_OPENCL_D3D11_NV
```

File: CMakeLists.txt [hunk 10]
```diff
@@ -365,6 +365,9 @@ OCV_OPTION(WITH_OPENCLAMDFFT "Include AMD OpenCL FFT library support" ON
 OCV_OPTION(WITH_OPENCLAMDBLAS "Include AMD OpenCL BLAS library support" ON
   VISIBLE_IF NOT ANDROID AND NOT IOS AND NOT WINRT
   VERIFY HAVE_CLAMDBLAS)
+  OCV_OPTION(WITH_OPENCLNVEXT "Include NVIDIA OpenCL Extensions support" OFF
+  VISIBLE_IF NOT ANDROID AND NOT IOS AND NOT WINRT
+  VERIFY HAVE_CLNVEXT)
```
Comment (alalek) [inline CMakeLists.txt]:
```
VISIBLE_IF WIN32 AND NOT WINRT
```

File: cmake/OpenCVDetectOpenCL.cmake [hunk 11]
```diff
@@ -36,7 +36,7 @@ if(OPENCL_FOUND)
     set(HAVE_OPENCL_SVM 1)
   endif()
 
-  set(OPENCL_INCLUDE_DIRS ${OPENCL_INCLUDE_DIR})
+  set(OPENCL_INCLUDE_DIRS ${OPENCL_INCLUDE_DIR} CACHE PATH "OpenCL include directory")
```
Comment (alalek) [inline cmake/OpenCVDetectOpenCL.cmake]: OPENCL_INCLUDE_DIRS should not be cached. Only `OPENCL_INCLUDE_DIR` is cached.

File: modules/core/src/directx.cpp (hunk 7 above)
Comment (Mainvooid) [reply modules/core/src/directx.cpp]: There's no problem. Both codes are there.

File: cmake/OpenCVDetectOpenCL.cmake (hunk 8 above)
Comment (Mainvooid) [reply cmake/OpenCVDetectOpenCL.cmake]: OK

File: cmake/templates/cvconfig.h.in (hunk 9 above)
Comment (Mainvooid) [reply cmake/templates/cvconfig.h.in]: changed to HAVE_OPENCL_D3D11_NV

File: cmake/OpenCVDetectOpenCL.cmake (hunk 11 above)
Comment (Mainvooid) [reply cmake/OpenCVDetectOpenCL.cmake]: OK

File: CMakeLists.txt (hunk 10 above)
Comment (Mainvooid) [reply CMakeLists.txt]: ok

File: cmake/OpenCVDetectOpenCL.cmake [hunk 12]
```diff
@@ -2,11 +2,26 @@ set(OPENCL_FOUND ON CACHE BOOL "OpenCL library is found")
 if(APPLE)
   set(OPENCL_LIBRARY "-framework OpenCL" CACHE STRING "OpenCL library")
   set(OPENCL_INCLUDE_DIR "" CACHE PATH "OpenCL include directory")
-else(APPLE)
+else()
   set(OPENCL_LIBRARY "" CACHE STRING "OpenCL library")
-  set(OPENCL_INCLUDE_DIR "${OpenCV_SOURCE_DIR}/3rdparty/include/opencl/1.2" CACHE PATH "OpenCL include directory")
+  if(WITH_OPENCL_D3D11_NV)
+    unset(OPENCL_INCLUDE_DIR CACHE)
```
Comment (alalek) [inline cmake/OpenCVDetectOpenCL.cmake]: Unsetting of cached variables is a bad practice. These variables are dedicated for configuration from "User".

You should check for `cl_d3d11_ext.h` in specified directory `OPENCL_INCLUDE_DIR`.
- found: enable feature: `set(HAVE_OPENCL_D3D11_NV ON)`
- not found: print warning, keep feature disabled.

File: modules/core/src/directx.cpp (hunk 4 above)
Comment (alalek) [reply modules/core/src/directx.cpp]: No need to do that.

Blank lines are useful for git merges to avoid unnecessary conflicts.

We don't use line fillers, like `////////////////////////////////////////` due different number of used symbols.

File: modules/core/src/directx.cpp [hunk 13]
```diff
@@ -678,36 +760,63 @@ Context& initializeContextFromDirect3DDevice9(IDirect3DDevice9* pDirect3DDevice9
...
     using namespace cv::ocl;
     static cl_platform_id initializedPlatform = NULL;
     cl_platform_id platform = (cl_platform_id)Platform::getDefault().ptr();
+    bool useCLNVEXT = false;
+
+#ifdef HAVE_OPENCL_D3D11_NV
     if (initializedPlatform != platform)
     {
-        clCreateFromD3D11Texture2DKHR = (clCreateFromD3D11Texture2DKHR_fn)
-                clGetExtensionFunctionAddressForPlatform(platform, "clCreateFromD3D11Texture2DKHR");
```
Comment (alalek) [inline modules/core/src/directx.cpp]: Consider keeping 8-space indentation in the list of parameters/arguments or continued declaration/statement.

Comment (Mainvooid) [reply modules/core/src/directx.cpp]: Done,I think this should be added to the Code Style Guide.

File: cmake/OpenCVDetectOpenCL.cmake (hunk 12 above)
Comment (Mainvooid) [reply cmake/OpenCVDetectOpenCL.cmake]: You're right.And if detection fails, it can also be searched from environment variables.

Comment (Mainvooid) [reply cmake/OpenCVDetectOpenCL.cmake]: Done

File: modules/core/src/directx.cpp (hunk 4 above)
Comment (Mainvooid) [reply modules/core/src/directx.cpp]: Okay

File: modules/core/src/directx.cpp (hunk 13 above)
Comment (Mainvooid) [reply modules/core/src/directx.cpp]: Done

File: cmake/OpenCVDetectOpenCL.cmake [hunk 14]
```diff
@@ -77,4 +96,5 @@ if(OPENCL_FOUND)
       list(APPEND OPENCL_INCLUDE_DIRS "${CLAMDBLAS_INCLUDE_DIR}")
     endif()
   endif()
+
```
Comment (alalek) [inline cmake/OpenCVDetectOpenCL.cmake]: Please remove unnecessary change.

File: cmake/OpenCVDetectOpenCL.cmake [hunk 15]
```diff
@@ -2,11 +2,30 @@ set(OPENCL_FOUND ON CACHE BOOL "OpenCL library is found")
...
+    if(NOT EXISTS "${OPENCL_INCLUDE_DIR}/CL/cl_d3d11_ext.h")
+      message(STATUS "CL/cl_d3d11_ext.h not found in OPENCL_INCLUDE_DIR. Automatic detection has been turned on")
+      unset(OPENCL_INCLUDE_DIR CACHE)
+      find_path(OPENCL_INCLUDE_DIR
+        NAMES CL/cl_d3d11_ext.h
+        PATHS ENV CUDA_PATH ENV ProgramFiles
+        PATH_SUFFIXES include
+        DOC "Extended opencl include directory")
+      if(OPENCL_INCLUDE_DIR)
+        set(HAVE_OPENCL_D3D11_NV on)
```
Comment (alalek) [inline cmake/OpenCVDetectOpenCL.cmake]: Please use uppercase "ON":
```
set(HAVE_OPENCL_D3D11_NV ON)
```

Comment (Mainvooid) [reply cmake/OpenCVDetectOpenCL.cmake]: Done.

**Comment by Mainvooid** [comment]
Here is a description about support for D3D9:
GTX Series graphics cards only TITAN X support `cl_nv_d3d9_sharing` extension.
I don't have the device, so I can't test it.

**Comment by Mainvooid** [comment]
I couldn't successfully rebase pull request from master to 3.4 branch.

**Comment by alalek** [comment]
@Mainvooid Please take a look on build errors here: http://pullrequest.opencv.org/#/summary/opencv

File: cmake/OpenCVDetectOpenCL.cmake [hunk 16]
```diff
@@ -2,11 +2,30 @@ set(OPENCL_FOUND ON CACHE BOOL "OpenCL library is found")
 if(APPLE)
   set(OPENCL_LIBRARY "-framework OpenCL" CACHE STRING "OpenCL library")
   set(OPENCL_INCLUDE_DIR "" CACHE PATH "OpenCL include directory")
-else(APPLE)
+else()
   set(OPENCL_LIBRARY "" CACHE STRING "OpenCL library")
-  set(OPENCL_INCLUDE_DIR "${OpenCV_SOURCE_DIR}/3rdparty/include/opencl/1.2" CACHE PATH "OpenCL include directory")
+  if(WITH_OPENCL_D3D11_NV)
+    if(NOT EXISTS "${OPENCL_INCLUDE_DIR}/CL/cl_d3d11_ext.h")
+      message(STATUS "CL/cl_d3d11_ext.h not found in OPENCL_INCLUDE_DIR. Automatic detection has been turned on")
```
Comment (mshabunin) [inline cmake/OpenCVDetectOpenCL.cmake]: Can we add this header file to `3rdparty/include/opencl`? Is it's license compatible? Is it standardized?

Comment (Mainvooid) [reply cmake/OpenCVDetectOpenCL.cmake]: Yes, it's compatible and standard. @alalek and I discussed earlier that it relies on `nvopencl64.dll` in the CUDA SDK, so it needs to be detected, and there is no need for HAVE_CUDA.

Comment (alalek) [reply cmake/OpenCVDetectOpenCL.cmake]: CUDA `.dll` (or `.lib`) is not needed to build OpenCV with support of this extension.

We need just this header file.
But I can't find proper legal source of this file (like [this](https://github.com/KhronosGroup/OpenCL-Headers/tree/master/CL)). Looks like this file comes from CUDA SDK only, under CUDA SDK's own license.

Comment (Mainvooid) [reply cmake/OpenCVDetectOpenCL.cmake]: Now there is a new problem, we still need to check whether the device has extension function at runtime, otherwise it will throw an uncapable `0xc0000005` exception when executing the `clGetDeviceIDsFromD3D11KHR` function that does not support extension.

> `clGetExtensionFunctionAddressForPlatform` does not guarantee that an extension function is 
actually supported by the platform. The application must also make a corresponding query
 using `clGetPlatformInfo (platform, CL_PLATFORM_EXTENSIONS, ... )` or `clGetDeviceInfo 
(device,CL_DEVICE_EXTENSIONS, ... )` to determine if an extension is supported by the OpenCL 
implementation.
**Notes**
Since there is no way to qualify the query with a device, the function pointer returned must work for all implementations of that extension on different devices for a platform. The behavior of calling a device extension function on a device not supporting that extension is undefined.

[clGetExtensionFunctionAddressForPlatform ](https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clGetExtensionFunctionAddressForPlatform.html)

**A recent commit has fixed this issue.**

---
The license is no different from other header files,and it has been declared at the beginning of the file.
```
* Copyright (c) 2008-2009 The Khronos Group Inc.
* Permission is hereby granted, free of charge...
```
So far, we can choose to add the header file or not, the runtime also needs to detect whether the device supports the corresponding extension.

But I think it's better to keep the status quo, because at compile time, users can clearly know that this extension needs CUDA support.

File: modules/core/src/directx.cpp [hunk 17]
```diff
@@ -260,71 +261,59 @@ Context& initializeContextFromD3D11Device(ID3D11Device* pD3D11Device)
     if (status != CL_SUCCESS)
         CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get number of platforms");
 
-    // TODO Filter platforms by name from OPENCV_OPENCL_DEVICE
+    size_t exts_len;
+    status = clGetPlatformInfo(platforms[0], CL_PLATFORM_EXTENSIONS, 0, NULL, &exts_len);
+    if (status != CL_SUCCESS)
+        CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get length of CL_PLATFORM_EXTENSIONS");
+    cv::AutoBuffer<char> extensions(exts_len);
+    status = clGetPlatformInfo(platforms[0], CL_PLATFORM_EXTENSIONS, exts_len, static_cast<void*>(extensions.data()), NULL);
```
Comment (alalek) [inline modules/core/src/directx.cpp]: > platforms[0]

I believe extensions check should be done in the loop below for each `platforms[i]`:
```
for (int i = 0; i < (int)numPlatforms; i++)
```

File: cmake/OpenCVDetectOpenCL.cmake (hunk 16 above)
Comment (alalek) [reply cmake/OpenCVDetectOpenCL.cmake]: If it has Khronos license lets add it near existed OpenCL header files (3rdparty/include/opencl/1.2/CL).
But keep CMake check for this file, because users can change headers sources via setting of own `OPENCL_INCLUDE_DIR`.

Also line
```
set(OPENCL_INCLUDE_DIR "${OpenCV_SOURCE_DIR}/3rdparty/include/opencl/1.2" CACHE PATH "OpenCL include directory")
```

should be moved before this check (check should be placed after "mark_as_advanced()" statement).

File: CMakeLists.txt [hunk 18]
```diff
@@ -384,6 +384,9 @@ OCV_OPTION(WITH_OPENCLAMDFFT "Include AMD OpenCL FFT library support" ON
 OCV_OPTION(WITH_OPENCLAMDBLAS "Include AMD OpenCL BLAS library support" ON
   VISIBLE_IF NOT ANDROID AND NOT IOS AND NOT WINRT
   VERIFY HAVE_CLAMDBLAS)
+OCV_OPTION(WITH_OPENCL_D3D11_NV "Include NVIDIA OpenCL D3D11 support" OFF
```
Comment (alalek) [inline CMakeLists.txt]: Lets try to enable this in default builds (to avoid breaking compilation of this code in the future):
`OFF` => `(WITH_DIRECTX)`

and move this option after `WITH_DIRECTX` declaration.

File: cmake/OpenCVDetectOpenCL.cmake (hunk 16 above)
Comment (Mainvooid) [reply cmake/OpenCVDetectOpenCL.cmake]: Done

File: cmake/OpenCVDetectOpenCL.cmake [hunk 19]
```diff
@@ -2,14 +2,33 @@ set(OPENCL_FOUND ON CACHE BOOL "OpenCL library is found")
...
-endif(APPLE)
+endif()
 mark_as_advanced(OPENCL_INCLUDE_DIR OPENCL_LIBRARY)
 
 if(OPENCL_FOUND)
+  if(WITH_OPENCL_D3D11_NV)
+    if(NOT EXISTS "${OPENCL_INCLUDE_DIR}/CL/cl_d3d11_ext.h")
+      message(STATUS "CL/cl_d3d11_ext.h not found in OPENCL_INCLUDE_DIR. Automatic detection has been turned on")
+      unset(OPENCL_INCLUDE_DIR CACHE)
+      find_path(OPENCL_INCLUDE_DIR
```
Comment (mshabunin) [inline cmake/OpenCVDetectOpenCL.cmake]: We should not unset/rewrite `OPENCL_INCLUDE_DIR` from cache, we should disable the feature in case when the header was not found:
```.cmake
if(WITH_OPENCL_D3D11_NV AND EXISTS "${OPENCL_INCLUDE_DIR}/CL/cl_d3d11_ext.h")
  set(HAVE_OPENCL_D3D11_NV ON)
endif()
```
User wishing to enable this feature should check cmake output or turn on `ENABLE_CONFIG_VERIFICATION` cmake option to verify it automatically.

Comment (Mainvooid) [reply cmake/OpenCVDetectOpenCL.cmake]: Since `cl_d3d11_ext.h` has been moved to the default directory, the search will no longer be necessary.Done.
